lib/cfg-grammar.y

```diff
@@ -925,14 +928,12 @@ template_content_list
 	;
 
 template_name_or_content
```

Comment:
Earlier I felt that template() options should define their templates locally, right where it is used. That's why the two use-cases started to drift away. Newer functions usually only allow to use the inline template syntax and did not support references.

The grammar has a conflict: "foobar" can both be a template name or an inline reference. This conflict is now resolved so that we look up if a string is matching a defined template and we assume it was meant to be a reference in this case.

With all this said, the question is if supporting named templates is something we want to restore throughout the codebase. This has been a compatibility feature for at least a decade.

But I can see people still using templates a lot (just look at SC4S for example). So I am less convinced by now that deprecating the support for named templates in newer modules was a good idea after all. 

Also, if this was a separate option (e.g. template-ref(<name>) and template(<inline>) I would be perfectly happy to support named templates throughout. The trickery we need to look up named templates is not something I am very fond of.

Maybe a good topic for discussion on Discord?

Reply:
lib/filter/filter-expr-grammar.ym:333:  | KW_TEMPLATE '(' template_content ')'
modules/afamqp/afamqp-grammar.ym:89:    | KW_BODY '(' template_content ')'         { afamqp_dd_set_body(last_driver, $3); }
? modules/affile/affile-grammar.ym:256: : template_content
? modules/affile/affile-grammar.ym:282: : template_content
modules/afsmtp/afsmtp-grammar.ym:86:        | KW_BODY '(' template_content ')'
modules/afstomp/afstomp-grammar.ym:76:        | KW_BODY '(' template_content ')'        { afstomp_dd_set_body(last_driver, $3); }
modules/correlation/correlation-grammar.ym:119:        | KW_MESSAGE_TEMPLATE '(' template_content ')'    { log_parser_set_template(last_parser, $3); }
modules/grpc/loki/loki-grammar.ym:101:  | KW_TEMPLATE '(' template_content ')' { loki_dd_set_message_template_ref(last_driver, $3); }
modules/http/http-grammar.ym:157:    | KW_BODY       '(' template_content ')'  { http_dd_set_body(last_driver, $3); log_template_unref($3); }
modules/kafka/kafka-grammar.ym:101:        | KW_MESSAGE '(' template_content ')'
modules/mqtt/mqtt-grammar.ym:102:        | KW_TEMPLATE '(' template_content ')'   { mqtt_dd_set_message_template_ref(last_driver, $3); }
modules/pseudofile/pseudofile-grammar.ym:77:    : KW_TEMPLATE '(' template_content ')'                  { pseudofile_dd_set_template(last_driver, $3); }
? rename KW_TEMPLATE to KW_KEY and should not become a reference modules/rate-limit-filter/rate-limit-grammar.ym:76:  : KW_TEMPLATE '(' template_content ')'

Reply:
We decided to do this in the places where bazsi pointed at. I will implement it in another PR.

Reply:
See #4679.